saw-core/src/SAWCore/Rewriter.hs

```diff
@@ -442,54 +465,49 @@ scEqsRewriteRules sc = mapM (scEqRewriteRule sc)
 -- * If the rhs is a record, then split into a separate rule for each accessor.
 scExpandRewriteRule :: SharedContext -> RewriteRule a -> IO (Maybe [RewriteRule a])
 scExpandRewriteRule sc (RewriteRule ctxt lhs rhs _ shallow ann) =
-  case rhs of
-    (R.asLambda -> Just (_, ty, body)) ->
-      do let ctxt' = ctxt ++ [ty]
-         lhs1 <- incVars sc 0 1 lhs
-         var0 <- scLocalVar sc 0
-         lhs' <- scApply sc lhs1 var0
-         return $ Just [mkRewriteRule ctxt' lhs' body shallow ann]
+  scAsLambda sc rhs >>= \case
+  Just (ec, body) ->
+    do let ctxt' = ctxt ++ [ec]
```

Comment:
...And as an added bonus, representing the context as an `IntSet` from the onset means that instead of appending things to the end of the context using `(++)` (which is O(n)), we can instead use `IntSet.insert` (which is O(log(n))).

Reply:
I want to keep the `ExtCns` variables as a list in the `ctxt` field of type `RewriteRule` in any case, because without it we don't have enough information to reconstruct the pi type for the theorem of the rewrite rule (e.g. in `propOfRewriteRule`). Having the bindings in order will also be important for type-checking rule instantiations in the presence of dependent types (which we don't do yet, but probably should).

We *could* add a field to `RewriteRule` that is a an `IntSet` precomputed from `ctxt`, if that's important for performance. But I'd prefer to implement optimizations like that only after properly investigating rewriter performance with the GHC profiler.

Reply:
Ah, I missed that we were making vital use of the list structure in `propOfRewriteRule`. In that case, feel free to ignore this suggestion.